src/autoscaler/metricsforwarder/server/custom_metrics_handlers.go

```diff
@@ -0,0 +1,206 @@
+				Message: "Basic authorization credential does not match"})
+			return
+		}
+	}
+
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		mh.logger.Error("error-reading-request-body", err, lager.Data{"body": r.Body})
+		handlers.WriteJSONResponse(w, http.StatusInternalServerError, models.ErrorResponse{
+			Code:    "Interal-Server-Error",
+			Message: "Error reading custom metrics request body"})
+		return
+	}
+	var metricsConsumer *models.MetricsConsumer
+	err = json.Unmarshal(body, &metricsConsumer)
+	if err != nil {
+		mh.logger.Error("error-unmarshaling-metrics", err, lager.Data{"body": r.Body})
+		handlers.WriteJSONResponse(w, http.StatusBadRequest, models.ErrorResponse{
+			Code:    "Bad-Request",
+			Message: "Error unmarshaling custom metrics request body"})
+		return
+	}
+	err = mh.validateCustomMetricTypes(appID, metricsConsumer)
+	if err != nil {
+		mh.logger.Error("failed-validating-metrictypes", err, lager.Data{"metrics": metricsConsumer})
+		handlers.WriteJSONResponse(w, http.StatusBadRequest, models.ErrorResponse{
+			Code:    "Bad-Request",
+			Message: "Error validating custom metrics type"})
+		return
+	}
+
+	metrics := mh.getMetrics(appID, metricsConsumer)
+
+	if len(metrics) <= 0 {
+		mh.logger.Debug("failed-parsing-custom-metrics-request-body", lager.Data{"metrics": metrics})
+		handlers.WriteJSONResponse(w, http.StatusBadRequest, models.ErrorResponse{
+			Code:    "Bad-Request",
+			Message: "Error parsing request body"})
+		return
+	}
+
+	mh.logger.Debug("custom-metrics-parsed-successfully", lager.Data{"metrics": metrics})
+	for _, metric := range metrics {
+		mh.metricForwarder.EmitMetric(metric)
+	}
+	w.WriteHeader(http.StatusOK)
+}
+
+func (mh *CustomMetricsHandler) validateCredentials(username string, usernameHash string, password string, passwordHash string) bool {
+	usernameAuthErr := bcrypt.CompareHashAndPassword([]byte(usernameHash), []byte(username))
+	passwordAuthErr := bcrypt.CompareHashAndPassword([]byte(passwordHash), []byte(password))
+	if usernameAuthErr == nil && passwordAuthErr == nil { // password matching successfull
+		return true
+	}
+	mh.logger.Debug("failed-to-authorize-credentials")
+	return false
+}
+
+func (mh *CustomMetricsHandler) validateCustomMetricTypes(appGUID string, metricsConsumer *models.MetricsConsumer) error {
+	standardMetricsTypes := make(map[string]struct{})
+	standardMetricsTypes[models.MetricNameMemoryUsed] = struct{}{}
+	standardMetricsTypes[models.MetricNameMemoryUtil] = struct{}{}
+	standardMetricsTypes[models.MetricNameThroughput] = struct{}{}
+	standardMetricsTypes[models.MetricNameResponseTime] = struct{}{}
+	standardMetricsTypes[models.MetricNameCPUUtil] = struct{}{}
+
+	allowedMetricTypeSet := make(map[string]struct{})
+	res, found := mh.allowedMetricCache.Get(appGUID)
+	if found {
+		// AllowedMetrics found in cache
+		allowedMetricTypeSet = res.(map[string]struct{})
+	} else {
+		//  AllowedMetrics not found in cache, find AllowedMetrics from Database
```

Comment:
again this will always querying database if the app keeps using wrong metric name